hrm-service/src/permissions/setupCanForRules.ts

```diff
@@ -56,28 +50,15 @@ const { Case, Contact, PostSurvey } = models;
     }
 
     return checkConditionsSets(conditionsState, conditionsSets);
-  });
+  };
 };
 
 export const setupCanForRules = (rules: RulesFile) => {
-  const cancan = new CanCan();
-  const { can, allow } = cancan;
-  const setupAllow = bindSetupAllow(allow);
-
-  // Configure Case permissions
-  Object.values(actionsMaps.case).forEach(action =>
-    setupAllow(User, action, Case, 'case', rules[action]),
-  );
-
-  // Configure Contact permissions
-  Object.values(actionsMaps.contact).forEach(action =>
-    setupAllow(User, action, Contact, 'contact', rules[action]),
-  );
-
-  // Configure PostSurvey permissions
-  Object.values(actionsMaps.postSurvey).forEach(action =>
-    setupAllow(User, action, PostSurvey, 'postSurvey', rules[action]),
-  );
+  const actionCheckers = Object.entries(actionsMaps).reduce<Record<Actions, ReturnType<typeof setupAllow>>>(
+    (outerAccum, [targetKind, actions]) => 
+      Object.entries(actions).reduce((innerAccum, [, action]) => ({ ...innerAccum, [action]: setupAllow(targetKind, rules[action]) }), outerAccum)
+    , null);
 
-  return can;
+  return (performer: User, action: Actions, target: any): boolean =>
+    actionCheckers[action](performer, target);
```

Comment:
Same idea as the original PR but this way we can add more `targetKind`s to `actionsMaps` and this code won't need update. Happy to be more explicit if this is ugly code tho.

Reply:
It's great that this is generalised, but my preference would be to try expressing the same generalised logic without resorting to reusing `reduce`. 'reduce' is difficult to read as a construct (especially nested `reduce`!) and if you can express the same logic just using map / filter / forEach etc. it's easier for others to read, even if it's more long winded. 

My preference would for `reduce` to only be used where absolutely necessary, i.e. where it would be a big performance hit to do it differently (and even then only if it's performance critical code).